### PR TITLE
U/danielsf/dc2/truth

### PR DESCRIPTION
**Pull request by danielsf**
This PR should add the reader for the DC2 truth catalog

### CODE REVIEW

**Comment by yymao** [comment]
Thanks, @danielsf! I am traveling this afternoon; will review later today or tomorrow.

**Comment by yymao** [comment]
@danielsf it seems to me that your implementation load the whole dataset into memory regardless which columns the user requests. Is this the intended behavior? Or should we execute the query within `dc2_truth_native_quantity_getter`?

**Comment by danielsf** [comment]
The only conscious choice was my desire to do the fastest, easiest thing possible.  I can move the query into the getter.

**Comment by danielsf** [comment]
The native_quantity_getter can only take one quantity name at a time, right?

I wonder if that will be less efficient, since we would have to execute one query per requested quantity, as opposed to just selecting all columns in one query.

**Comment by yymao** [comment]
@danielsf In fact, you can just overwrite this method as the following: (I haven't tested it but think it should work)

```python
def _load_quantities(self, quantities, native_quantity_getter):
    native_quantities = tuple(self._translate_quantities(quantities))
    native_data = dict(zip(native_quantities, native_quantity_getter(native_quantities)))
    return {q: self._assemble_quantity(q, native_data) for q in quantities}
```

**Comment by danielsf** [comment]
@yymao 
Your scheme does work.  I have pushed the appropriate changes.  Thanks!

File: GCRCatalogs/dc2_truth.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    def _iter_native_dataset(self, native_filters=None):
+        cursor = self._conn.cursor()
+
+        query_where_clause = ' FROM truth'
+
+        if native_filters is not None:
+            query_where_clause += ' WHERE'
+
+            if not isinstance(native_filters, list) and not isinstance(native_filters, tuple):
```
Comment (yymao) [inline GCRCatalogs/dc2_truth.py]: The user should provide `native_filters` as a list or tuple. So I don't think this is necessary (or should raise error instead?)

File: GCRCatalogs/dc2_truth.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                native_filters = [native_filters]
+
+            for i_filt, filt in enumerate(native_filters):
```
Comment (yymao) [inline GCRCatalogs/dc2_truth.py]: No sure if I understand the logic here. In this case, `native_filters` should be a list of strings. So I think we just need
```python
query_where_clause += ' AND '.join(native_filters)
```

File: GCRCatalogs/dc2_truth.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                if i_filt > 0:
+                    query_where_clause += ' AND'
+                query_where_clause += ' {}'.format(filt)
+
+        # define a method to return a native_quantity_getter
+        # with the API expected by the GCR
+        def dc2_truth_native_quantity_getter(quantities):
+            query = 'SELECT'
+            for i_qty, qty in enumerate(quantities):
```
Comment (yymao) [inline GCRCatalogs/dc2_truth.py]: Line 60 -- 66, do
```python
query = 'SELECT {} {}'.format(','.join(quantities), query_where_clause)
```

File: GCRCatalogs/dc2_truth.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+from GCR import BaseGenericCatalog
+
+__all__ = ["DC2TruthCatalogReader"]
+
+
+class DC2TruthCatalogReader(BaseGenericCatalog):
+
+    def _subclass_init(self, **kwargs):
+
+        self._allow_string_native_filter = True
```
[the remark this answers is not in the document]
Comment (danielsf) [reply GCRCatalogs/dc2_truth.py]: Done

File: GCRCatalogs/dc2_truth.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+        if not os.path.isfile(kwargs['filename']):
+            raise ValueError("%s is not a valid filename" % kwargs['filename'])
+
+        self._conn = sqlite3.connect(kwargs['filename'])
+
+        # get the descriptions of the columns as provided in the sqlite database
+        cursor = self._conn.cursor()
+        metadata = cursor.execute('SELECT name, description FROM column_descriptions').fetchall()
+        self._column_descriptions = {}
```
[the remark this answers is not in the document]
Comment (danielsf) [reply GCRCatalogs/dc2_truth.py]: Done

File: GCRCatalogs/dc2_truth.py (hunk 2 above)
Comment (danielsf) [reply GCRCatalogs/dc2_truth.py]: The logic here accomplishes the same thing, but I guess it doesn't matter how we do it.

File: GCRCatalogs/dc2_truth.py (hunk 3 above)
Comment (danielsf) [reply GCRCatalogs/dc2_truth.py]: Done

**Comment by yymao** [comment]
the catalog name is also a bit confusing. How about `dc2_truth_protoDC2` ?

**Comment by danielsf** [comment]
I thought it would make sense to have the catalog name mirror the protoDC2 extragalactic catalogs.

We should probably include something in the name to indicate this is a truth catalog for run 1.1, not run 1.2.

I guess if we specify the run, we do not need to specify that it is protoDC2, rather than cosmoDC2.

How about `dc2_truth_run1.1`.  I will add a line to the catalog description indicating which version of `protoDC2` the catalog corresponds to.

**Comment by katrinheitmann** [comment]
I think that name captures the important information!

**Comment by yymao** [comment]
I think we should avoid using `proto-dc2` at the beginning because those are all extragalactic catalogs.

**Comment by danielsf** [comment]
Ready for re-review

**Comment by yymao** [comment]
@danielsf all looks good. I'd like to merge this with #148 as both requires a newer version of GCR. So I'll wait until someone reviews #148.

**Comment by danielsf** [comment]
Sounds good.  Push the button whenever you're ready.

**Pull request merged**